## confirm missing values

In [72]:
%matplotlib inline
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import scipy.stats as st
import sklearn.linear_model as linear_model
import seaborn as sns
from sklearn.model_selection import KFold
from IPython.display import HTML, display
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

train = pd.read_csv('C:/Users/dohan/Desktop/house_prj/house-price-project/data/train.csv')
test = pd.read_csv('C:/Users/dohan/Desktop/house_prj/house-price-project/data/test.csv')

In [3]:
total = train.isnull().sum().sort_values(ascending=False)
percent = (train.isnull().sum()/train.isnull().count()).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(20)

Total   Percent
PoolQC         1453  0.995205
MiscFeature    1406  0.963014
Alley          1369  0.937671
Fence          1179  0.807534
FireplaceQu     690  0.472603
LotFrontage     259  0.177397
GarageCond       81  0.055479
GarageType       81  0.055479
GarageYrBlt      81  0.055479
GarageFinish     81  0.055479
GarageQual       81  0.055479
BsmtExposure     38  0.026027
BsmtFinType2     38  0.026027
BsmtFinType1     37  0.025342
BsmtCond         37  0.025342
BsmtQual         37  0.025342
MasVnrArea        8  0.005479
MasVnrType        8  0.005479
Electrical        1  0.000685
Utilities         0  0.000000

## 1. GarageX (GarageCond, GarageType, GarageYrBlt, GarageFinish, GarageQual)

### 결측값 판단

차고가 없는 집을 NaN으로 처리. 결측값은 없다.

In [8]:
garage_cols = ['GarageYrBlt','GarageCond','GarageFinish','GarageQual','GarageType']
train[garage_cols][train['GarageYrBlt'].isnull()]

GarageYrBlt GarageCond GarageFinish GarageQual GarageType
39            NaN        NaN          NaN        NaN        NaN
48            NaN        NaN          NaN        NaN        NaN
78            NaN        NaN          NaN        NaN        NaN
88            NaN        NaN          NaN        NaN        NaN
89            NaN        NaN          NaN        NaN        NaN
99            NaN        NaN          NaN        NaN        NaN
108           NaN        NaN          NaN        NaN        NaN
125           NaN        NaN          NaN        NaN        NaN
127           NaN        NaN          NaN        NaN        NaN
140           NaN        NaN          NaN        NaN        NaN
148           NaN        NaN          NaN        NaN        NaN
155           NaN        NaN          NaN        NaN        NaN
163           NaN        NaN          NaN        NaN        NaN
165           NaN        NaN          NaN        NaN        NaN
198           NaN        NaN          NaN        NaN        NaN
210           NaN        NaN          NaN        NaN        NaN
241           NaN        NaN          NaN        NaN        NaN
250           NaN        NaN          NaN        NaN        NaN
287           NaN        NaN          NaN        NaN        NaN
291           NaN        NaN          NaN        NaN        NaN
307           NaN        NaN          NaN        NaN        NaN
375           NaN        NaN          NaN        NaN        NaN
386           NaN        NaN          NaN        NaN        NaN
393           NaN        NaN          NaN        NaN        NaN
431           NaN        NaN          NaN        NaN        NaN
434           NaN        NaN          NaN        NaN        NaN
441           NaN        NaN          NaN        NaN        NaN
464           NaN        NaN          NaN        NaN        NaN
495           NaN        NaN          NaN        NaN        NaN
520           NaN        NaN          NaN        NaN        NaN
...           ...        ...          ...        ...        ...
954           NaN        NaN          NaN        NaN        NaN
960           NaN        NaN          NaN        NaN        NaN
968           NaN        NaN          NaN        NaN        NaN
970           NaN        NaN          NaN        NaN        NaN
976           NaN        NaN          NaN        NaN        NaN
1009          NaN        NaN          NaN        NaN        NaN
1011          NaN        NaN          NaN        NaN        NaN
1030          NaN        NaN          NaN        NaN        NaN
1038          NaN        NaN          NaN        NaN        NaN
1096          NaN        NaN          NaN        NaN        NaN
1123          NaN        NaN          NaN        NaN        NaN
1131          NaN        NaN          NaN        NaN        NaN
1137          NaN        NaN          NaN        NaN        NaN
1143          NaN        NaN          NaN        NaN        NaN
1173          NaN        NaN          NaN        NaN        NaN
1179          NaN        NaN          NaN        NaN        NaN
1218          NaN        NaN          NaN        NaN        NaN
1219          NaN        NaN          NaN        NaN        NaN
1234          NaN        NaN          NaN        NaN        NaN
1257          NaN        NaN          NaN        NaN        NaN
1283          NaN        NaN          NaN        NaN        NaN
1323          NaN        NaN          NaN        NaN        NaN
1325          NaN        NaN          NaN        NaN        NaN
1326          NaN        NaN          NaN        NaN        NaN
1337          NaN        NaN          NaN        NaN        NaN
1349          NaN        NaN          NaN        NaN        NaN
1407          NaN        NaN          NaN        NaN        NaN
1449          NaN        NaN          NaN        NaN        NaN
1450          NaN        NaN          NaN        NaN        NaN
1453          NaN        NaN          NaN        NaN        NaN

[81 rows x 5 columns]

### 처리 방안
Nan을 None으로 변경한다.

In [9]:
for cols in garage_cols:
    if train[cols].dtype==np.object:
        train.loc[train[cols].isnull(), cols] = 'None'
    else:
        train.loc[train[cols].isnull(), cols] = 0

## 2. LotFrontage


### 결측값 판단
LotArea, LotConfig, LotShape에는 missing value가 없다. 하지만 LotFrontage에 259개의 missing value가 존재한다. 따라서 이는 결측값이다.

### 처리방안 1
LotFrontage의 median 값을 활용

In [63]:
frontage_median = train['LotFrontage'].median()
train['LotFrontage'].fillna(frontage_median, inplace=True)
train[['LotFrontage', 'LotArea']]

LotFrontage  LotArea
0            65.0     8450
1            80.0     9600
2            68.0    11250
3            60.0     9550
4            84.0    14260
5            85.0    14115
6            75.0    10084
7            69.0    10382
8            51.0     6120
9            50.0     7420
10           70.0    11200
11           85.0    11924
12           69.0    12968
13           91.0    10652
14           69.0    10920
15           51.0     6120
16           69.0    11241
17           72.0    10791
18           66.0    13695
19           70.0     7560
20          101.0    14215
21           57.0     7449
22           75.0     9742
23           44.0     4224
24           69.0     8246
25          110.0    14230
26           60.0     7200
27           98.0    11478
28           47.0    16321
29           60.0     6324
...           ...      ...
1430         60.0    21930
1431         69.0     4928
1432         60.0    10800
1433         93.0    10261
1434         80.0    17400
1435         80.0     8400
1436         60.0     9000
1437         96.0    12444
1438         90.0     7407
1439         80.0    11584
1440         79.0    11526
1441         69.0     4426
1442         85.0    11003
1443         69.0     8854
1444         63.0     8500
1445         70.0     8400
1446         69.0    26142
1447         80.0    10000
1448         70.0    11767
1449         21.0     1533
1450         60.0     9000
1451         78.0     9262
1452         35.0     3675
1453         90.0    17217
1454         62.0     7500
1455         62.0     7917
1456         85.0    13175
1457         66.0     9042
1458         68.0     9717
1459         75.0     9937

[1460 rows x 2 columns]

### 처리방안 2
LotFrontage는 집터와 접한 땅의 길이이다. LotArea는 집터 면적이다.
따라서 LotArea가 클수록 LotFrontage값도 클 것이라 가정하고, LotArea의 sqare root 값을 LotFrontage의 missing value에 넣어준다.

In [75]:
isnullLotFrontage = train["LotFrontage"].isnull()
sqrtLotArea = np.sqrt(train.loc[isnullLotFrontage].LotArea)
train.loc[isnullLotFrontage, "LotFrontage"] = np.sqrt(train.loc[isnullLotFrontage, "LotArea"])
train[['LotFrontage', 'LotArea']]

LotFrontage  LotArea
0       65.000000     8450
1       80.000000     9600
2       68.000000    11250
3       60.000000     9550
4       84.000000    14260
5       85.000000    14115
6       75.000000    10084
7      101.892100    10382
8       51.000000     6120
9       50.000000     7420
10      70.000000    11200
11      85.000000    11924
12     113.877127    12968
13      91.000000    10652
14     104.498804    10920
15      51.000000     6120
16     106.023582    11241
17      72.000000    10791
18      66.000000    13695
19      70.000000     7560
20     101.000000    14215
21      57.000000     7449
22      75.000000     9742
23      44.000000     4224
24      90.807489     8246
25     110.000000    14230
26      60.000000     7200
27      98.000000    11478
28      47.000000    16321
29      60.000000     6324
...           ...      ...
1430    60.000000    21930
1431    70.199715     4928
1432    60.000000    10800
1433    93.000000    10261
1434    80.000000    17400
1435    80.000000     8400
1436    60.000000     9000
1437    96.000000    12444
1438    90.000000     7407
1439    80.000000    11584
1440    79.000000    11526
1441    66.528190     4426
1442    85.000000    11003
1443    94.095696     8854
1444    63.000000     8500
1445    70.000000     8400
1446   161.684879    26142
1447    80.000000    10000
1448    70.000000    11767
1449    21.000000     1533
1450    60.000000     9000
1451    78.000000     9262
1452    35.000000     3675
1453    90.000000    17217
1454    62.000000     7500
1455    62.000000     7917
1456    85.000000    13175
1457    66.000000     9042
1458    68.000000     9717
1459    75.000000     9937

[1460 rows x 2 columns]